src/apps/vpn/models/servercity.cpp

```diff
@@ -167,15 +150,70 @@ int ServerCity::connectionScore() const {
 
 unsigned int ServerCity::latency() const {
   ServerLatency* serverLatency = MozillaVPN::instance()->serverLatency();
-  qint64 now = QDateTime::currentSecsSinceEpoch();
-  int activeServerCount = 0;
+  int numLatencySamples = 0;
   uint32_t sumLatencyMsec = 0;
   for (const QString& pubkey : m_servers) {
-    if (serverLatency->getCooldown(pubkey) <= now) {
-      sumLatencyMsec += serverLatency->getLatency(pubkey);
-      activeServerCount++;
+    unsigned int rtt = serverLatency->getLatency(pubkey);
+    if (rtt > 0) {
+      sumLatencyMsec += rtt;
+      numLatencySamples++;
     }
   }
 
-  return (sumLatencyMsec + activeServerCount - 1) / activeServerCount;
+  if (numLatencySamples == 0) {
+    return 0;
+  }
+  return (sumLatencyMsec + numLatencySamples - 1) / numLatencySamples;
+}
+
+int ServerCity::connectionScore() const {
+  int score = baseCityScore(MozillaVPN::instance()->location()->countryCode());
+  if (score <= Unavailable) {
+    return score;
+  }
+
+  // If there is no latency data, then we haven't actually measured anything
+  // and have no connectivity signals to report.
+  unsigned int cityLatencyMsec = latency();
+  if (cityLatencyMsec == 0) {
+    return NoData;
+  }
+
+  // Increase the score if the location has better than average latency.
+  if (cityLatencyMsec < MozillaVPN::instance()->serverLatency()->avgLatency()) {
+    score++;
+    // Give the location another point if the latency is *very* fast.
+    if (cityLatencyMsec < SCORE_EXCELLENT_LATENCY_THRESHOLD) {
+      score++;
+    }
+  }
+
+  if (score > Excellent) {
+    score = Excellent;
+  }
+  return score;
+}
+
+int ServerCity::multiHopScore(const QString& country,
```

Comment:
Seems like this and the above method would also benefit from tests. I'm not sure what our mocking capabilities are in order to be able to do this well. 

In particular, if I'm understanding the PR message correctly, given that we won't be using this code for a bit (...maybe never??) having tests will help us come back to it later and document then intended scoring.

Reply:
Yes, unit tests are good, I will add some.

This method is invoked for the multihop exit servers, so it is used whenever the feature is enabled.

Reply:
Hmm, the dependency tangle between `ServerCity` and `ServerLatency` classes is turning out to make this a pain in the butt to mock out for unit testing, but I will get this tested or die trying.

Reply:
In case it helps, I am a proponent of (re-)architecting code for testability.